packages/opentelemetry-cloud-trace-propagator/src/CloudPropagator.ts

```diff
@@ -0,0 +1,89 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {
+  HttpTextPropagator,
+  Context,
+  SetterFunction,
+  GetterFunction,
+  TraceFlags,
+} from '@opentelemetry/api';
+import { decToHex, hexToDec } from 'hex2dec';
+import {
+  setExtractedSpanContext,
+  getParentSpanContext,
+  randomSpanId,
+  randomTraceId,
+} from '@opentelemetry/core';
+
+/**
+ * This file implements propagation for the Stackdriver Trace v1 Trace Context
+ * format.
+ *
+ * The header specification is:
+ * "X-Cloud-Trace-Context: TRACE_ID/SPAN_ID;o=TRACE_TRUE"
+ * Where:
+ *  {TRACE_ID} is a 32-character hexadecimal value representing a 128-bit
+ *    number. It should be unique between your requests, unless you
+ *    intentionally want to bundle the requests together.
+ *  {SPAN_ID} is the decimal representation of the (unsigned) span ID. It
+ *    should be 0 for the first span in your trace. For subsequent requests,
```

Comment:
@mayurkale22 I don't know how this (`should be 0 for the first span in your trace`) is going to play with the [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/c33919ea1265081967287a3c3160ba823797bbce/specification/trace/api.md#spancontext) requiring the `spanId` to have at least one non-zero byte

Reply:
I think this is not the right statement, AFAIK it is not mentioned in the official doc: https://cloud.google.com/trace/docs/setup#force-trace

Reply:
Great, the docs should be updated to reflect this.

Also, are those the official docs? Because in this package from Google https://googleapis.github.io/google-cloud-dotnet/docs/Google.Cloud.Diagnostics.AspNetCore/api/Google.Cloud.Diagnostics.Common.TraceHeaderContext.html#properties, source code: https://github.com/googleapis/google-cloud-dotnet/blob/master/apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/Trace/TraceHeaderContext.cs#L47

They accept the trace flag to have values in range [0-3], and traceId can be with either lowercase or uppercase chars.

There's also this package https://www.npmjs.com/package/gcloud-trace, from which the current implementation is derived, claiming that the `spanId` is optional or even the pair `traceId/spanId` if the `traceFlag` is set.

Maybe an excerpt from internal docs (proto definition) should be used to describe the header.

Reply:
SpanIDs should never be zero. The API docs do specify that: https://cloud.google.com/trace/docs/reference/v2/rpc/google.devtools.cloudtrace.v2?hl=en#google.devtools.cloudtrace.v2.Span

@sk- The reason this header may contain flags in the range 0-3 is that bit 1 of the header controls stack trace collection, which is supported by the Cloud Trace product. This is not yet supported by OpenTelemetry, but it may be adopted in the future: https://github.com/open-telemetry/oteps/pull/69.
For OpenTelemetry interoperability we have to strip bit 1 when ingesting this header.

Reply:
The linked gcloud-trace npm package is a 3rd party package. A missing span ID can never be used as an incoming trace context. The only use-case for missing span IDs is when this header is used as a response header for signalling downstream sampling decisions. This feature may be implemented in the future in OpenTelemetry using the traceresponse header: https://github.com/w3c/trace-context/blob/master/spec/21-http_response_header_format.md